experiments/compositions/composition/tests/kind/cluster.go

```diff
@@ -0,0 +1,235 @@
+	err := c.Delete()
+	if err != nil {
+		return err
+	}
+	clusterConfig, err := c.kindClusterDefinition()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(clusterConfig)
+
+	c.config, err = c.createCluster(clusterConfig)
+
+	kindClusterSet.available.Store(c.Name(), c)
+	return nil
+}
+
+// Config return rest.Config
+func (c *kindCluster) Config() *rest.Config {
+	return c.config
+}
+
+// Name return name
+func (c *kindCluster) Name() string {
+	return c.name
+}
+
+func (c *kindCluster) String() string {
+	return c.name
+}
+
+// LoadImage loads a docker image into the cluster
+func (c *kindCluster) LoadImage(image string) error {
+	err := exec.Command("kind", "load", "docker-image", image, "--name", c.name).Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete deletes all clusters
+func (c *kindCluster) Delete() error {
+	kindClusterSet.available.Delete(c.Name())
+	err := exec.Command("kind", "delete", "cluster", "--name", c.name).Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Exists checks if the cluster exists
+func (c *kindCluster) Exists() (bool, error) {
+	output, err := exec.Command("kind", "get", "clusters").CombinedOutput()
+	if err != nil {
+		return false, err
+	}
+
+	for _, cluster := range strings.Split(string(output), "\n") {
+		if cluster == c.name {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *kindCluster) kindClusterDefinition() (string, error) {
+	ipAddress, err := c.getHostIPAddress()
+	if err != nil {
+		return "", err
+	}
+	clusterConfigFile, err := os.CreateTemp("", "kind-cluster.yaml")
+	if err != nil {
+		return "", err
+	}
+	defer clusterConfigFile.Close()
+	kindClusterConfig := `kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  # Allow connections to the API Sever with the CloudTop IP address
+  apiServerAddress: "` + ipAddress + `"`
+	bytes := []byte(kindClusterConfig)
+	_, err = clusterConfigFile.Write(bytes)
+	return clusterConfigFile.Name(), err
+}
+
+func (c *kindCluster) createCluster(clusterConfig string) (*rest.Config, error) {
+	_, err := exec.Command("kind", "create", "cluster", "--name", c.name, "--config", clusterConfig).CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfigFile, err := os.CreateTemp("", "kubeconfig.yaml")
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(kubeConfigFile.Name())
+	content, err := exec.Command("kind", "get", "kubeconfig", "--name", c.name).CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+	bytes := []byte(content)
+	_, err = kubeConfigFile.Write(bytes)
+	kubeConfigFile.Close()
+
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeConfigFile.Name()},
+		&clientcmd.ConfigOverrides{
+			ClusterInfo: clientcmdapi.Cluster{
+				Server: "",
+			},
+			CurrentContext: "",
+		}).ClientConfig()
+}
+
+func (c *kindCluster) getHostIPAddress() (string, error) {
+	conn, err := net.Dial("udp", "8.8.8.8:80")
```

Comment:
What is this magic?   This deserves a comment!

Reply:
some hack i found somewhere on the internet before gemini existed.